galsim/wfirst/wfirst_wcs.py

```diff
@@ -0,0 +1,571 @@
+        # easier to follow.
+        alpha1 = crval1
+        delta1 = crval2
+        alpha2 = alpha1
+        # Increment dec by half-way from center to edge of the SCA.
+        # This seems a reasonable compromise in making the step big enough to avoid round-off errors
+        # becoming too large and avoiding non-linearities in the trigonometric functions.
+        delta2 = delta1 + float(galsim.wfirst.n_pix) * pix_scale / 4
+
+        # If these two points straddle dec = 90, shift points the other direction.
+        if delta2/galsim.degrees > 90.:
+            delta2 = delta1
+            delta1 = delta2 - float(galsim.wfirst.n_pix) * pix_scale / 4
+        # Only need to check North pole: if near South pole, the step above always takes the second
+        # point farther from the pole.
+
+        # Project onto tangent plane.
+        sin_dec1 = np.sin(delta1.rad())
+        cos_dec1 = np.cos(delta1.rad())
+        sin_dec2 = np.sin(delta2.rad())
+        cos_dec2 = np.cos(delta2.rad())
+        cos_dalpha1 = np.cos((alpha1 - use_ra).rad())
+        sin_dalpha1 = np.sin((alpha1 - use_ra).rad())
+        cos_dalpha2 = cos_dalpha1
+        sin_dalpha2 = sin_dalpha1
+
+        # Can compute sin_theta1 and sin_theta2 directly without first computing theta1:
+        sin_theta1 = sin_dec1*sin_decp + cos_dec1*cos_decp*cos_dalpha1
+        sin_theta2 = sin_dec2*sin_decp + cos_dec2*cos_decp*cos_dalpha2
+        # These should always be in range 0-1, but look out for roundoff error:
+        if sin_theta1 < 0.: sin_theta1 = 0.
+        if sin_theta1 > 1.: sin_theta1 = 1.
+        if sin_theta2 < 0.: sin_theta2 = 0.
+        if sin_theta2 > 1.: sin_theta2 = 1.
+        # This is always in the range 0-90, so can compute cos_theta simply:
+        cos_theta1 = np.sqrt(1.-sin_theta1*sin_theta1)
+        cos_theta2 = np.sqrt(1.-sin_theta2*sin_theta2)
+
+        phi1 = phi_p + \
+            np.arctan2(-cos_dec1*sin_dalpha1, 
+                        sin_dec1*cos_decp-cos_dec1*sin_decp*cos_dalpha1)*galsim.radians
+        phi2 = phi_p + \
+            np.arctan2(-cos_dec2*sin_dalpha2,
+                        sin_dec2*cos_decp-cos_dec2*sin_decp*cos_dalpha2)*galsim.radians
+
+        # Convert to X-Y positions:
+        # Don't need to protect against divide by zero as theta is always near 90 degrees.
+        xtp1 = cos_theta1*np.sin(phi1.rad())*galsim.radians/sin_theta1
+        ytp1 = -cos_theta1*np.cos(phi1.rad())*galsim.radians/sin_theta1
+        xtp2 = cos_theta2*np.sin(phi2.rad())*galsim.radians/sin_theta2
+        ytp2 = -cos_theta2*np.cos(phi2.rad())*galsim.radians/sin_theta2
+        dxtp = xtp2 - xtp1
+        dytp = ytp2 - ytp1
+
+        # Finally have the ingredients for computing the position angle of this SCA Y axis.
+        pa_sca = np.arctan2(-cos_sca_rot*dxtp.rad()+sin_sca_rot*dytp.rad(),
+                             sin_sca_rot*dxtp.rad()+cos_sca_rot*dytp.rad())*galsim.radians
```

Comment:
I just pushed a commit that includes a great deal of simplifying (and more explaining) of the WCS code.  I think it's now much clearer, but please see what you think.